### PR TITLE
add support for sql query timeout

### PR DESCRIPTION
**Pull request by mauri870**
For #1657

### CODE REVIEW

**Comment by otoolep** [comment]
So, what was the secret?

File: DOC/DATA_API.md [hunk 1]
```diff
@@ -199,6 +199,14 @@ curl -XPOST 'localhost:4001/db/execute?timeout=2m' -H "Content-Type: application
 ]'
 ```
 
+### SQL Query Timeout
```
Comment (otoolep) [inline DOC/DATA_API.md:202]: I don't maintain these docs any longer. Instead we will update the docs on https://github.com/rqlite/rqlite.io/blob/master/content/en/docs/API/api/_index.md

File: db/db_test.go [hunk 2]
```diff
@@ -986,6 +988,117 @@ func Test_ParallelOperationsInMemory(t *testing.T) {
 	exWg.Wait()
 }
 
+func mustSetupDbForTimeoutTests(t *testing.T, n int) (*DB, string) {
```
Comment (otoolep) [inline db/db_test.go:991]: Nit: Go style says to use "DB", not "Db".

File: db/db_test.go [hunk 3]
```diff
@@ -986,6 +988,117 @@ func Test_ParallelOperationsInMemory(t *testing.T) {
...
+func Test_ExecShouldTimeout(t *testing.T) {
+	db, path := mustSetupDbForTimeoutTests(t, 1000)
+	defer db.Close()
+	defer os.Remove(path)
+
+	q := `
+INSERT INTO test_table (key1, key_id, key2, key3, key4, key5, key6, data)
+SELECT t1.key1 || t2.key1, t1.key_id || t2.key_id, t1.key2 || t2.key2, t1.key3 || t2.key3, t1.key4 || t2.key4, t1.key5 || t2.key5, t1.key6 || t2.key6, t1.data || t2.data
+FROM test_table t1 LEFT OUTER JOIN test_table t2`
+	mustTimeoutExecute(db, q, 1*time.Millisecond)
```
Comment (otoolep) [inline db/db_test.go:1047]: Is this a test? Are you testing that it does timeout? Then you should check the error code returned by the underlying Execute or Timeout function.  Implicitly testing by not panic'ing isn't the right approach.

The "must" functions are for setup purposes, for doing stuff in tests that is not meant to be tested, and must execute for the actual testing to commence. `mustSetupDbForTimeoutTests` is a good example of such a function.

File: db/db_test.go [hunk 4]
```diff
@@ -986,6 +988,117 @@ func Test_ParallelOperationsInMemory(t *testing.T) {
...
+
+func Test_QueryShouldTimeout(t *testing.T) {
+	db, path := mustSetupDbForTimeoutTests(t, 1000)
+	defer db.Close()
+	defer os.Remove(path)
+
+	q := `SELECT key1, key_id, key2, key3, key4, key5, key6, data
+	FROM test_table
+	ORDER BY key2 ASC`
+	mustTimeoutQuery(db, q, 1*time.Microsecond)
```
Comment (otoolep) [inline db/db_test.go:1066]: Ditto, this is not the right way to test this. Do the query, and test that the error that comes back looks good.

File: http/query_params.go [hunk 5]
```diff
@@ -123,6 +123,16 @@ func (qp QueryParams) Key() string {
 	return qp["key"]
 }
 
+// SqlTimeout returns the value of the key named "sql_timeout_seconds".
+func (qp QueryParams) SqlTimeout(def time.Duration) time.Duration {
```
Comment (otoolep) [inline http/query_params.go:127]: Go Style says to call this "SQLTimeout"

Comment (otoolep) [reply http/query_params.go:127]: Also, add a check at line 29 for the validity of this variable, if exists, then you can parse directly here.

**Comment by mauri870** [comment]
Basically what you said, the query was executing too fast. I hope the tests are not flaky, I ran with -count=1000 to make sure and it seems to have consistent results.

File: http/query_params.go (hunk 5 above)
Comment (otoolep) [reply http/query_params.go:127]: I'm also torn about the use of "sql". I think "db_timeout" is better.

**Review by otoolep — COMMENTED** [review]
Thank you, this looks generally like what I would do.

File: http/query_params.go (hunk 5 above)
Comment (mauri870) [reply http/query_params.go:127]: Thank you; I was quite unsure about the naming here. I wanted something that conveys that this is a timeout per query, not per transaction or per database connection.

File: DOC/DATA_API.md (hunk 1 above)
Comment (mauri870) [reply DOC/DATA_API.md:202]: Thanks, I moved the docs to https://github.com/rqlite/rqlite.io/pull/10

File: db/db_test.go (hunk 3 above)
Comment (mauri870) [reply db/db_test.go]: Agreed, updated the tests :)

**Comment by otoolep** [comment]
Thanks again, looks good, will make a couple of follow-up changes.

**Pull request merged**